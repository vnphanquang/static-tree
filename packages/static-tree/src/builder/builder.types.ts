/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { TNode, TNodeData, TNodeInit } from '../node';

import { ExtendedTNodeBuilder } from './builder';

/**
 * @public
 *
 * {@link TNode} extended with children inline as properties, intended to provide
 * idiomatic dot notation parent-to-child node access, so that you can do:
 *
 * ```typescript
 * root.child.grandChild.$.path(); // -\> `root/child/grandChild`
 * ```
 *
 * instead of:
 *
 * ```typescript
 * root.$.children().find(c => c.key === 'child')?.$.children().find(c => c.key === 'grandChild')?.$.path();
 * ```
 *
 * @remarks
 *
 * {@link ExtendedTNode} is constructed behind the scene using {@link tBuild} (preferred)
 * and {@link ExtendedTNodeBuilder}.
 */
export type ExtendedTNode<
  Key extends string = never,
  AncestorKeys extends string = never,
  ChildrenRecord extends Record<string, ExtendedTNode> = {},
  Data extends TNodeData = {},
> = TNode<Data, Key, AncestorKeys> & ChildrenRecord;

/**
 * @public
 *
 * Instruction for building nested {@link ExtendedTNode}.
 *
 */
export interface ExtendedTNodeBuildConfig<
  Key extends string,
  AncestorKeys extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData = {},
> {
  pathResolver?: TNodeInit<Data>['pathResolver'];
  build?: ExtendedTNodeBuildCallback<Key, AncestorKeys, ChildrenRecord, Data>;
}

/**
 * @public
 *
 * Build callback for adding data & nested children
 */
export type ExtendedTNodeBuildCallback<
  Key extends string,
  AncestorKeys extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData = {},
> = (builder: ExtendedTNodeBuilder<Key, AncestorKeys>) => ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, Data>;

/**
 * @public
 *
 * Output of {@link tBuild}
 */
export interface TBuildOutput<
  Key extends string,
  AncestorKeys extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
> {
  /**
   * built node with type safety for children and data
   */
  node: ExtendedTNode<Key, AncestorKeys, ChildrenRecord, Data>;
  /**
   * builder for advanced use cases. See {@link tBuild} for examples
   */
  builder: ExtendedTNodeBuilder<Key, AncestorKeys, ChildrenRecord, Data>;
}
