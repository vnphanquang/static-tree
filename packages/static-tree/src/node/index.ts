export * from './node';
export * from './node.types';
