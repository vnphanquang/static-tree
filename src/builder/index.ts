export * from './builder';
export * from './builder.types';
