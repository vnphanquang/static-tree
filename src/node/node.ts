/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type {
  MinimalSerializedTNode,
  SerializedTNode,
  TNodeGetPathParams,
  TNodeData,
  TNodeInit,
  TNodeSerializeOptions,
  VerboseSerializedTNode,
  TNodePublicApi,
  TNodeInternalApi,
} from './node.types';

/**
 * @public
 *
 * Implementation for each node of the static tree
 *
 * @remarks
 *
 * ### Constructing TNode
 *
 * `TNode` is exported in the public api but you should probably avoid constructing
 * `TNode` directly. Instead, use the {@link tBuild} helper, as they provide
 * type safety and extension to `TNode` for better developer experience.
 *
 * ### Using TNode
 *
 * By design, `TNode` exposes two proxy method collections through two getters:
 *
 * - `$`: intended for public usage. You should interact with TNode through this.
 * - `__`: intended for internal usage only. Avoid this.
 *
 * ```typescript
 * const key = node.$.key();
 * const path = node.$.path({ depth: 2, separator: '.' });
 * // ...
 * ```
 *
 * @example
 *
 * ```typescript
 * import { TNode } from 'static-tree';
 *
 * // only do this only if you know exactly what you want
 * const parent = new TNode('parent');
 * const data = { prop: 'value' }
 * const child = new TNode('child', { parent, data });
 * ```
 */
export class TNode<Data extends TNodeData = {}> {
  private _parent: TNode | null;
  private _key: string;
  private _depth: number;
  private _pathSegments: string[];
  private _children: TNode[];
  private _data: Data;

  /**
   * @param key - key of this node
   * @param options - the {@link TNodeInit} instruction
   */
  constructor(key: string, options: TNodeInit<Data> = {}) {
    const { parent, data } = options;
    this._children = [];
    this._key = key;
    this._data = data ?? ({} as any);
    this.__setParent(parent);
  }

  /**
   * Construct a TNode and all its children from a serialized object (most
   * likely from the `TNode.$.serialize` method)
   *
   * @param serialized - a {@link SerializedTNode} object
   * @param parent - a {@link TNode} parent to attach to
   * @returns node {@link TNode}
   */
  public static from(serialized: SerializedTNode, parent?: TNode): TNode<any> {
    const { key, children, data } = serialized;
    const node = new TNode(key, { parent, data });
    for (const child of children) {
      const childNode = TNode.from(child, node);
      node.__addChildren(childNode);
    }
    return node;
  }

  /**
   * Change parent of current node and handle side effects, including:
   *
   * - remove current node from parent's children (if parent exists)
   *
   * - recompute path segments & depth
   *
   * @param parent - a {@link TNode} to attach to
   * @returns this {@link TNode}
   */
  private __setParent(parent?: TNode): TNode<Data> {
    const oldParent = this._parent;

    if (oldParent && oldParent.$children().includes(this)) {
      oldParent.__removeChildren(this);
    }

    this._parent = parent ?? null;

    if (this._parent && !this._parent.$children().includes(this)) {
      this._parent.$children().push(this);
    }

    this._depth = (this._parent?.$depth() ?? 0) + 1;

    this._pathSegments = [this.$key()];
    let traversedParent = this._parent;
    while (traversedParent) {
      this._pathSegments.unshift(traversedParent.$key());
      traversedParent = traversedParent.$parent();
    }

    for (const child of this.$children()) {
      child.__setParent(this);
    }

    return this;
  }

  /**
   * Add children to record & set their parent to current node
   *
   * @param nodes - {@link TNode}[] list of nodes to add
   * @returns this {@link TNode}
   */
  private __addChildren(...nodes: TNode[]): TNode<Data> {
    this.$children().push(...nodes);
    for (const node of nodes) {
      node.__setParent(this);
    }
    return this;
  }

  /**
   * Remove specified children from this node, and set their parent to null
   *
   * @param nodes - {@link TNode}[] list of nodes to remove
   * @returns this {@link TNode}
   */
  private __removeChildren(...nodes: TNode[]): TNode<Data> {
    this._children = this._children.filter((n) => nodes.every((r) => r != n));
    for (const node of nodes) {
      node.__setParent(null);
    }
    return this;
  }

  /**
   * Set internal data for this node
   *
   * @param data - {@link TNodeData} to add
   * @returns this {@link TNode}
   */
  private __setData<NewData extends TNodeData>(data: NewData): TNode<NewData> {
    this._data = data as any;
    return this as TNode<NewData>;
  }

  /**
   * @returns key of this node
   */
  private $key(): string {
    return this._key;
  }

  /**
   * @returns depth of this node
   */
  private $depth(): number {
    return this._depth;
  }

  /**
   * @returns parent of this node, or null if node is a root
   */
  private $parent(): TNode | null {
    return this._parent;
  }

  /**
   * @returns children {@link TNode} of this node
   */
  private $children(): TNode[] {
    return this._children;
  }

  /**
   * Construct the paths for this node
   *
   * @param params - a {@link TNodeGetPathParams} instruction for how path is constructed
   * @returns path
   */
  private $path(params: TNodeGetPathParams = {}): string {
    const { depth = this._pathSegments.length, separator = '/', reversed = false, till } = params;
    const len = this._pathSegments.length;

    let count = Math.abs(depth);
    let node: TNode = this as TNode;
    let anchor = len;
    while (count > 0 && node) {
      count--;
      anchor--;
      if (node == till) {
        break;
      }
      node = node.$parent();
    }

    let start = anchor;
    let end = len;
    if (depth < 0) {
      start = 0;
      end = till ? len - anchor : depth;
    }

    let segments = this._pathSegments.slice(start, end);
    if (reversed) segments = segments.reverse();

    return segments.join(separator);
  }

  /**
   * @returns whether node is a root (no parent)
   */
  private $isRoot(): boolean {
    return !this.$parent();
  }

  /**
   * @returns root {@link TNode} of the tree
   */
  private $root(): TNode {
    let node: TNode = this as TNode;
    while (node.$parent()) {
      node = node.$parent();
    }
    return node;
  }

  /**
   * @param options - a {@link TNodeSerializeOptions} object
   * @returns a serializable {@link SerializedTNode} object
   */
  private $serialize<M extends boolean>(
    options: TNodeSerializeOptions<M, Data> = {},
  ): M extends true ? VerboseSerializedTNode : MinimalSerializedTNode {
    const { verbose, dataSerializer } = options;
    let data: any = this._data;
    if (typeof dataSerializer === 'boolean' && dataSerializer === false) {
      data = {};
    } else if (typeof dataSerializer === 'function') {
      data = dataSerializer(this);
    }
    return {
      key: this.$key(),
      children: this.$children().map((c) => c.$serialize(options)),
      data,
      ...(verbose && {
        depth: this.$depth(),
        isRoot: this.$isRoot(),
        path: this.$path(),
        pathSegments: this._pathSegments,
      }),
    } as M extends true ? VerboseSerializedTNode : MinimalSerializedTNode;
  }

  /**
   * @returns internal data {@link TNodeData} of this node
   */
  private $data(): Data {
    return this._data;
  }

  /**
   * @returns the {@link TNodePublicApi} of {@link TNode}
   */
  public get $(): TNodePublicApi<Data> {
    return {
      key: this.$key.bind(this),
      depth: this.$depth.bind(this),
      path: this.$path.bind(this),
      root: this.$root.bind(this),
      isRoot: this.$isRoot.bind(this),
      children: this.$children.bind(this),
      parent: this.$parent.bind(this),
      serialize: this.$serialize.bind(this),
      data: this.$data.bind(this),
    };
  }

  /**
   * @returns the {@link TNodeInternalApi} of {@link TNode}
   */
  public get __(): TNodeInternalApi<Data> {
    return {
      setParent: this.__setParent.bind(this) as typeof this.__setParent,
      addChildren: this.__addChildren.bind(this) as typeof this.__addChildren,
      removeChildren: this.__removeChildren.bind(this) as typeof this.__removeChildren,
      setData: this.__setData.bind(this) as typeof this.__setData,
    };
  }
}
