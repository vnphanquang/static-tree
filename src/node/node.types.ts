/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { TNode } from './node';

export interface TNodeInit<D extends Record<string, any> = {}> {
  parent?: TNode;
  data?: D;
}

export interface TNodeGetPathParams {
  separator?: string;
  depth?: number;
}

type DeepRecord<T> = {
  [key: string]: T | DeepRecord<T>;
};

export type Serializable = DeepRecord<
  string | number | boolean | boolean | Array<Serializable> | ReadonlyArray<Serializable>
>;

export interface TNodeSerializeOptions<V extends boolean> {
  dataSerializer?: false | (<D extends Record<string, any> = {}>(node: TNode<D>) => Serializable);
  verbose?: V;
}

export interface MinimalSerializedTNode {
  key: string;
  children: MinimalSerializedTNode[];
  data: Serializable;
}

export interface VerboseSerializedTNode extends MinimalSerializedTNode {
  children: VerboseSerializedTNode[];
  pathSegments: string[];
  depth: number;
  isRoot: boolean;
  path: string;
}

export type SerializedTNode = VerboseSerializedTNode | MinimalSerializedTNode;

export type TNodeData = Record<string, any>;
