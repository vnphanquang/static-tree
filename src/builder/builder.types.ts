/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import type { TNode, TNodeData } from '../node';

import { ExtendedTNodeBuilder } from './builder';

/**
 * @public
 *
 * {@link TNode} extended with children as properties
 */
export type ExtendedTNode<
  ChildrenRecord extends Record<string, ExtendedTNode> = {},
  Data extends TNodeData = {},
> = TNode<Data> & ChildrenRecord;

/**
 * @public
 *
 * Instruction for building nested {@link ExtendedTNode}.
 *
 * Can be one of the following:
 *
 * - a string as the key of the node to be built
 *
 * - an {@link ExtendedTNodeBuilder} object
 *
 * - an object with `key` & an optional {@link ExtendedTNodeBuildCallback}
 *
 */
export type ExtendedTNodeBuildInput<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData = {},
> =
  | Key
  | ExtendedTNodeBuilder<Key, ChildrenRecord, Data>
  | {
      /**
       * key for {@link ExtendedTNode} to be built
       */
      key: Key;
      build?: ExtendedTNodeBuildCallback<Key, ChildrenRecord, Data>;
    };

/**
 * @public
 *
 * Build callback for adding data & nested children
 */
export type ExtendedTNodeBuildCallback<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData = {},
> = (builder: ExtendedTNodeBuilder<Key>) => ExtendedTNodeBuilder<Key, ChildrenRecord, Data>;

/**
 * @public
 *
 * Output of {@link tBuild}
 */
export interface TBuildOutput<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
> {
  /**
   * built node with type safety for children and data
   */
  node: ExtendedTNode<ChildrenRecord, Data>;
  /**
   * builder for advanced use cases. See {@link tBuild} for examples
   */
  builder: ExtendedTNodeBuilder<Key, ChildrenRecord, Data>;
}
