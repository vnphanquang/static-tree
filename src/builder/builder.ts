/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { TNode } from '../node';
import type { TNodeData, TNodeInit } from '../node';

import type { ExtendedTNode, ExtendedTNodeBuildInput, TBuildOutput } from './builder.types';

/**
 * @public
 *
 * Builder for {@link ExtendedTNode}
 *
 * @remarks
 *
 * This is used internally by {@link tBuild}. Alternatively, this can
 * be used if a more method-based, verbose solution is preferred.
 *
 * @example
 *
 * ```typescript
 * import { ExtendedTNodeBuilder } from 'static-tree';
 *
 * const node = new ExtendedTNodeBuilder('key')
 *   .addData({ number: 1, boolean: true, string: 'string' })
 *   .addChild({
 *      key: 'childOne',
 *      build: (builder) => builder
 *        .addData({ childData: 'something else' })
 *        .addChildren('grandChild),
 *   })
 *   .addChild('childTwo', builder => builder.addChild('grandChild'))
 *   .build();
 * ```
 *
 * @example
 *
 * For advanced usage, `ExtendedTNodeBuilder` can also be used
 * as the input in `addChild`, should it provide any benefit.
 *
 * ```typescript
 * import { tBuild, ExtendedTNodeBuilder } from 'static-tree';
 *
 * const externalBuilder = new ExtendedTNodeBuilder('external')
 *  .addData({ some: 'some' })
 *  .addChild('someChild');
 *
 * const { node } = tBuild({
 *   key: 'root',
 *   build: (builder) => builder.addChild(externalBuilder),
 * });
 *
 * node.external.someChild.$.path(); // -\> `root/external/someChild`
 * ```
 */
export class ExtendedTNodeBuilder<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode> = {},
  Data extends TNodeData = {},
> {
  private _node: ExtendedTNode;
  private _key: Key;

  /**
   * @param key - key for the {@link ExtendedTNode} to be built
   * @param parent - parent {@link TNode} to attach to
   */
  constructor(key: Key, options: Omit<TNodeInit<Data>, 'data'> = {}) {
    this._node = new TNode(key, options);
    this._key = key;
  }

  /**
   * add a type-safe child to the {@link ExtendedTNode} to be built
   *
   * @example
   *
   * `addChild` with just a key (no nested children, no internal data)
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder.addChild('child'),
   * });
   * ```
   *
   * @example
   *
   * `addChild` with a build callback for nested children & adding data
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder
   *     .addChild({
   *        key: 'child',
   *        build: (builder) => builder
   *          .addData({ some: 'child data' })
   *          .addChild('grandChildren'),
   *     }),
   * });
   * ```
   *
   * @example
   *
   * `addChild` with external `ExtendedTNodeBuilder` (advanced use case)
   *
   * ```typescript
   * import { tBuild, ExtendedTNodeBuilder } from 'static-tree';
   *
   * const externalBuilder = new ExtendedTNodeBuilder('external');
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder.addChildren(externalBuilder),
   * });
   * ```
   *
   * @param input - an {@link ExtendedTNodeBuildInput} instruction
   * @returns this {@link ExtendedTNodeBuilder}
   */
  public addChild<
    ChildKey extends string,
    GrandChildrenRecord extends Record<string, ExtendedTNode>,
    ChildData extends TNodeData = {},
  >(
    input: ExtendedTNodeBuildInput<ChildKey, GrandChildrenRecord, ChildData>,
  ): ExtendedTNodeBuilder<
    Key,
    ChildrenRecord & Record<ChildKey, ExtendedTNode<GrandChildrenRecord, ChildData>>,
    Data
  > {
    let key: ChildKey;
    let node: TNode;

    if (typeof input === 'string') {
      key = input;
      node = new ExtendedTNodeBuilder(key, { parent: this._node }).build();
    } else if (input instanceof ExtendedTNodeBuilder) {
      key = input._key;
      node = input.build();
      node.__.setParent(this._node);
    } else {
      key = input.key;
      node =
        input
          .build?.(
            new ExtendedTNodeBuilder(key, {
              parent: this._node,
              pathResolver: input.pathResolver,
            }),
          )
          .build() ?? new TNode(key, { parent: this._node });
    }

    (this._node as any)[key] = node;

    return this as unknown as ExtendedTNodeBuilder<
      Key,
      ChildrenRecord & Record<ChildKey, ExtendedTNode<GrandChildrenRecord, ChildData>>,
      Data
    >;
  }

  /**
   * add type-safe data to the {@link ExtendedTNode} to be built
   *
   * @example
   *
   * Using `tBuild`
   *
   * ```typescript
   * import { tBuild } from 'static-tree';
   *
   * const { node } = tBuild({
   *   key: 'root',
   *   build: (builder) => builder.addData({ some: 'data' }),
   * });
   * ```
   *
   * @example
   *
   * Using `ExtendedTNodeBuilder`
   *
   * @param data - the {@link TNodeData}
   * @returns this {@link ExtendedTNodeBuilder}
   */
  public addData<D extends TNodeData>(data: D): ExtendedTNodeBuilder<Key, ChildrenRecord, D> {
    this._node.__.setData(data);
    return this as ExtendedTNodeBuilder<Key, ChildrenRecord, D>;
  }

  /**
   * Build the {@link ExtendedTNode}
   *
   * @returns node {@link ExtendedTNode}
   */
  public build(): ExtendedTNode<ChildrenRecord, Data> {
    return this._node as ExtendedTNode<ChildrenRecord, Data>;
  }
}

/**
 * @public
 *
 * Build helper that uses {@link ExtendedTNodeBuilder} internally.
 *
 * @remarks
 *
 * Along with a built {@link ExtendedTNode}, `tBuild` outputs
 * an {@link ExtendedTNodeBuilder} for advanced use cases, see examples
 * for more details.
 *
 * @example
 *
 * Typical usage
 *
 * ```typescript
 * import { tBuild } = 'static-tree';
 *
 * const { node } = tBuild({
 *  key: 'root',
 *  build: (builder) => builder
 *    .addData({ some: 'root data' })
 *    .addChild('leaf')
 *    .addChild({
 *      key: 'child',
 *      build: (builder) => builder.addChild({
 *        key: 'grandChild',
 *        build: (builder) => builder.addData({ some: 'grand child data' }),
 *      }),
 *    }),
 * });
 *
 * node.$.data().some // -\> 'root data';
 * node.child.grandChild.$.data().some // -\> 'grand child data';
 * node.child.grandChild.$.path() // -\> 'root/child/grandChild';
 * ```
 *
 * @example
 *
 * Use `tBuild` builder output in nested `ExtendedTNodeBuildCallback`
 *
 * ```typescript
 * const { builder: externalBuilder } = tBuild({
 *   key: 'external',
 *   build: (builder) => builder.addChild('child').addData({ some: 'data' }),
 * });
 *
 * const { node } = tBuild({
 *   key: 'root',
 *   build: (builder) => builder.addChild(externalBuilder),
 * });
 * ```
 *
 * @param build - build instruction
 * @param key - key for the {@link ExtendedTNode} to be built
 * @returns
 */
export function tBuild<
  Key extends string,
  ChildrenRecord extends Record<string, ExtendedTNode>,
  Data extends TNodeData,
>(
  input: ExtendedTNodeBuildInput<Key, ChildrenRecord, Data>,
): TBuildOutput<Key, ChildrenRecord, Data> {
  let builder: ExtendedTNodeBuilder<Key, ChildrenRecord, Data>;

  if (typeof input === 'string') {
    builder = new ExtendedTNodeBuilder(input);
  } else if (input instanceof ExtendedTNodeBuilder) {
    builder = input;
  } else {
    builder =
      input.build?.(
        new ExtendedTNodeBuilder(input.key, {
          pathResolver: input.pathResolver,
        }),
      ) ?? new ExtendedTNodeBuilder(input.key);
  }

  return {
    builder,
    node: builder.build(),
  };
}
