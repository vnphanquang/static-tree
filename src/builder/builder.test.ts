import { test, expect, describe } from 'vitest';

import { TNode } from '../node';

import { ExtendedTNodeBuilder, tBuild } from './builder';

export const TNODE_MOCK_DATA = {
  number: 1,
  string: 'string',
  boolean: true,
  object: {
    nested: 'nested',
  },
  array: [
    {
      id: 'test',
    },
  ],
  specials: {
    date: new Date('2022-08-18'),
    node: new TNode('example'),
    map: new Map(),
    set: new Set(),
    function: () => 2,
  },
} as const;

const { node: sTree } = tBuild({
  key: 'key',
  build: (builder) =>
    builder.addData(TNODE_MOCK_DATA).addChild({
      key: 'child',
      build: (builder) =>
        builder
          .addChild('justKey')
          .addChild({ key: 'inputJustKey' })
          .addChild(new ExtendedTNodeBuilder('external'))
          .addChild({
            key: 'tBuild',
            build: (builder) =>
              builder
                .addChild(tBuild({ key: 'tBuildInputJustKey' }).builder)
                .addChild(tBuild('tBuildJustKey').builder)
                .addChild(tBuild(new ExtendedTNodeBuilder('tBuildWithBuilder')).builder)
                .addChild(
                  tBuild({
                    key: 'tBuildFull',
                    build: (builder) => builder.addChild('final'),
                  }).builder,
                ),
          }),
    }),
});

describe('tBuild', () => {
  test('data should be typed and returned correctly', () => {
    expect(sTree.$.data()).toMatchObject(TNODE_MOCK_DATA);
    expect(sTree.$.data().number).toBe(TNODE_MOCK_DATA.number);
  });

  test('children should be typed and match', () => {
    expect(sTree.$.key()).toBe('key');
    expect(sTree.child.$.path()).toBe('key/child');
    expect(sTree.child.justKey.$.path()).toBe('key/child/justKey');
    expect(sTree.child.inputJustKey.$.path()).toBe('key/child/inputJustKey');
    expect(sTree.child.external.$.path()).toBe('key/child/external');
    expect(sTree.child.tBuild.$.path()).toBe('key/child/tBuild');
    expect(sTree.child.tBuild.tBuildInputJustKey.$.path()).toBe(
      'key/child/tBuild/tBuildInputJustKey',
    );
    expect(sTree.child.tBuild.tBuildJustKey.$.path()).toBe('key/child/tBuild/tBuildJustKey');
    expect(sTree.child.tBuild.tBuildWithBuilder.$.path()).toBe(
      'key/child/tBuild/tBuildWithBuilder',
    );
    expect(sTree.child.tBuild.tBuildFull.$.path()).toBe('key/child/tBuild/tBuildFull');
    expect(sTree.child.tBuild.tBuildFull.final.$.path()).toBe('key/child/tBuild/tBuildFull/final');
  });

  test('serialization should match', () => {
    expect(sTree.$.serialize({ verbose: true })).toMatchSnapshot();
  });
});
