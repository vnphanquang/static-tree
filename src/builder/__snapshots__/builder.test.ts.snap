// Vitest Snapshot v1

exports[`tBuild > serialization should match 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "data": {},
          "depth": 3,
          "isRoot": false,
          "key": "justKey",
          "path": "key/child/justKey",
          "pathSegments": [
            "key",
            "child",
            "justKey",
          ],
        },
        {
          "children": [],
          "data": {},
          "depth": 3,
          "isRoot": false,
          "key": "inputJustKey",
          "path": "key/child/inputJustKey",
          "pathSegments": [
            "key",
            "child",
            "inputJustKey",
          ],
        },
        {
          "children": [],
          "data": {},
          "depth": 3,
          "isRoot": false,
          "key": "external",
          "path": "key/child/external",
          "pathSegments": [
            "key",
            "child",
            "external",
          ],
        },
        {
          "children": [
            {
              "children": [],
              "data": {},
              "depth": 4,
              "isRoot": false,
              "key": "tBuildInputJustKey",
              "path": "key/child/tBuild/tBuildInputJustKey",
              "pathSegments": [
                "key",
                "child",
                "tBuild",
                "tBuildInputJustKey",
              ],
            },
            {
              "children": [],
              "data": {},
              "depth": 4,
              "isRoot": false,
              "key": "tBuildJustKey",
              "path": "key/child/tBuild/tBuildJustKey",
              "pathSegments": [
                "key",
                "child",
                "tBuild",
                "tBuildJustKey",
              ],
            },
            {
              "children": [],
              "data": {},
              "depth": 4,
              "isRoot": false,
              "key": "tBuildWithBuilder",
              "path": "key/child/tBuild/tBuildWithBuilder",
              "pathSegments": [
                "key",
                "child",
                "tBuild",
                "tBuildWithBuilder",
              ],
            },
            {
              "children": [
                {
                  "children": [],
                  "data": {},
                  "depth": 5,
                  "isRoot": false,
                  "key": "final",
                  "path": "key/child/tBuild/tBuildFull/final",
                  "pathSegments": [
                    "key",
                    "child",
                    "tBuild",
                    "tBuildFull",
                    "final",
                  ],
                },
              ],
              "data": {},
              "depth": 4,
              "isRoot": false,
              "key": "tBuildFull",
              "path": "key/child/tBuild/tBuildFull",
              "pathSegments": [
                "key",
                "child",
                "tBuild",
                "tBuildFull",
              ],
            },
          ],
          "data": {},
          "depth": 3,
          "isRoot": false,
          "key": "tBuild",
          "path": "key/child/tBuild",
          "pathSegments": [
            "key",
            "child",
            "tBuild",
          ],
        },
      ],
      "data": {},
      "depth": 2,
      "isRoot": false,
      "key": "child",
      "path": "key/child",
      "pathSegments": [
        "key",
        "child",
      ],
    },
  ],
  "data": {
    "array": [
      {
        "id": "test",
      },
    ],
    "boolean": true,
    "number": 1,
    "object": {
      "nested": "nested",
    },
    "specials": {
      "date": 2022-08-18T00:00:00.000Z,
      "function": [Function],
      "map": Map {},
      "node": TNode {
        "_children": [],
        "_data": {},
        "_depth": 1,
        "_key": "example",
        "_parent": null,
        "_pathSegments": [
          "example",
        ],
      },
      "set": Set {},
    },
    "string": "string",
  },
  "depth": 1,
  "isRoot": true,
  "key": "key",
  "path": "key",
  "pathSegments": [
    "key",
  ],
}
`;
